File: in-order-traversal/in-order.pl
% In Order Traversal in Prolog

% Base case: empty tree returns empty list
inorder(nil, []).

% Recursive case: traverse left, current node, then right
inorder(node(Left, Value, Right), Result) :-
    inorder(Left, LeftResult),        % Get left subtree result
    inorder(Right, RightResult),      % Get right subtree result
    append(LeftResult, [Value], TempResult),      % Add current node after left
    append(TempResult, RightResult, Result).      % Add right subtree last

% Example tree structure: node(Left, Value, Right)
% Tree: 1 with children 2(4,5) and 3(6)
% Expected result: [4, 2, 5, 1, 3, 6]

% run this query to see result
% inorder(node(node(node(nil, 4, nil), 2, node(nil, 5, nil)), 1, node(nil, 3, node(nil, 6, nil))), Result).

File: in-order-traversal/InOrder.java
// In-order traversal implementation
public class InOrder {
    // Recursive pre-order traversal with printing
    void printInorder(Node node) {
        if (node == null)               // Base case: null node
            return;

        printInorder(node.left);      // Traverse left subtree
        System.out.print(node.data + " ");  // Print current node
        printInorder(node.right);     // Traverse right subtree
    }

    // Main method creates tree and executes traversal
    public static void main(String[] args) {
        // Build tree: 1 with children 2(4,5) and 3(6) 
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.right.right = new Node(6);
        
        // Output: 4 2 5 1 3 6
        InOrder tree = new InOrder();
        tree.printInorder(root);
    }
}